### PR TITLE
Allow using PORO decorators

### LINKED ISSUE #6248: Allow using PORO decorators
## Motivation

We were removing Draper from our projects template in favor of PORO presenter since we were filling decorators were used in a none-optimal way, plus we wanted to have fewer dependencies and a lower learning curve for newcomers. During that process, we found out how often we used decorators on ActiveAdmin pages and how useful they could be.

We also found that ActiveAdmin has a hard dependency on Draper making it very complex to develop presenters/decorators that conform with the expected API (this issue was very helpful https://github.com/activeadmin/activeadmin/issues/3600)

## Implementation

I think most of the issues with decorators and the need to delegate to the underlying ActiveRecord Relation come from decorating in an early stage. Since the decorator's methods are meant to be used just at the presentation layer I believe we can decorate the collection closer to the point where the methods will be used and just work with the plain ActiveRecord Relation before that point.

I'll open a PR with this implementation change, I'm still trying to figure out a better way to get the decorating method to the Index Page but it's more to check out if it's something desirable. I tested it locally and automated tests are green but let me know if there's anything else I can do to assure this works properly!

### PR DESCRIPTION
**Pull request by brunvez**
Closes #6248

### CODE REVIEW

**Comment by brunvez** [comment]
@deivid-rodriguez @javierjulio any thoughts on this? It'd really help our team to break the dependency with Draper

**Comment by deivid-rodriguez** [comment]
@brunvez Thanks for this! :tada:

 'm currently working on upgrading the whole development & runtime stack to ruby 2.7. That's alomost done now. After that I'll work on #6115, and this one is next on my list.

**Comment by brunvez** [comment]
Hey @deivid-rodriguez sorry to poke again but I was wondering if there's anything I can do to move this PR forward 😄

**Comment by deivid-rodriguez** [comment]
Hei @brunvez! 

Since #6115 will require a bit more work, I'll skip directly to reviewing this PR. Hope to do it very soon!

File: lib/active_admin/resource_controller/decorators.rb [hunk 1]
```diff
@@ -77,9 +77,9 @@ def self.wrap!(parent, name)
         # Draper::CollectionDecorator was introduced in 1.0.0
         # Draper::Decorator#collection_decorator_class was introduced in 1.3.0
         def self.find_collection_decorator(decorator)
-          if Dependency.draper? '>= 1.3.0'
+          if Dependency.draper?('>= 1.3.0') && decorator <= Draper::Decorator
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource_controller/decorators.rb:80]: I removed all this stuff because we don't really need it. Can you rebase and adapt the PR? I guess it's a matter of saying

```suggestion
          if Dependency.draper? && decorator <= Draper::Decorator
```

and restoring the final branch.

Comment (brunvez) [reply lib/active_admin/resource_controller/decorators.rb]: There we go, by removing the if here the expected interface for the decorator class changed to also require `collection_decorator_class`. But I think is a good tradeoff to have the code here cleaner.

Comment (deivid-rodriguez) [reply lib/active_admin/resource_controller/decorators.rb]: I wasn't meaning to suggest changing the interface for PORO decorators. I just meant to adapt the code to the lack of versions checks for `draper, with:

```ruby
def self.find_collection_decorator(decorator)
  if Dependency.draper? && decorator <= Draper::Decorator
    decorator.collection_decorator_class
  else
    decorator
end
```

Do you think we should require PORO decorators to define `collection_decorator_class`?

By the way, this remainded me that we need docs for this feature at `docs/11-decorators.md`.

Comment (deivid-rodriguez) [reply lib/active_admin/resource_controller/decorators.rb]: I proposed #6351. With that PR I believe we don't need any changes here in this file, nor to add a requirement for PORO decorators to define `collection_decorator_class`.

Comment (brunvez) [reply lib/active_admin/resource_controller/decorators.rb]: Cool, yeah on a second thought I think making PORO define the `collection_decorator_class` to just return `self` might not be too elegant. Do you want to merge that refactor first and I'll add docs on how to use PORO presenters?

Comment (deivid-rodriguez) [reply lib/active_admin/resource_controller/decorators.rb]: Yeah! :+1:

Comment (javierjulio) [reply lib/active_admin/resource_controller/decorators.rb]: @brunvez we just merged in the change so you are good to rebase. Thanks!

File: docs/11-decorators.md [hunk 2]
```diff
@@ -42,6 +37,33 @@ ActiveAdmin.register Post do
...
+  end
+
+  def self.decorate(record_or_collection, *_)
+    if record_or_collection.is_a?(Enumerable)
+      record_or_collection.map { |individual_record| new(individual_record) }
+    else
+      new(record_or_collection)
+    end
+  end
+end
```
Comment (deivid-rodriguez) [inline docs/11-decorators.md:61]: Something I don't love about this is that each decorator needs to add this code, which is as you point out later very generic. I feel ActiveAdmin should just work without having to add this. So I took a stab at it by copying the basic functionality from draper's collection presenter. I came up with this:

```diff
--- a/lib/active_admin/resource_controller/decorators.rb
+++ b/lib/active_admin/resource_controller/decorators.rb
@@ -48,18 +48,38 @@ module ActiveAdmin
         end
       end
 
+      class CollectionDecorator
+        # @return the collection being decorated.
+        attr_reader :object
+
+        # @return [Class] the decorator class used to decorate each item, as set by {#initialize}.
+        attr_reader :decorator_class
+
+        array_methods = Array.instance_methods - Object.instance_methods
+        delegate :==, :as_json, *array_methods, to: :decorated_collection
+
+        def initialize(object, with:)
+          @object = object
+          @decorator_class = with
+        end
+
+        class << self
+          alias_method :decorate, :new
+        end
+
+        def decorated_collection
+          @decorated_collection ||= object.map { |item| decorator_class.new(item) }
+        end
+      end
+
       class Wrapper
         @cache = {}
 
         def self.wrap(decorator)
-          collection_decorator = find_draper_collection_decorator(decorator)
+          collection_decorator = find_draper_collection_decorator(decorator) || CollectionDecorator
 
-          if collection_decorator
-            name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
-            @cache[name] ||= wrap! collection_decorator, name
-          else
-            decorator
-          end
+          name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
+          @cache[name] ||= wrap! collection_decorator, name
         end
 
         private
```

With that in place, I believe we don't require any other changes whatsoever to support PORO decorators, and people can just plugin their own decorators, regardless of whether they are draper decorators or other types.

What do you think?

Comment (brunvez) [reply docs/11-decorators.md:61]: I like the idea of allowing people to use decorators without needing to add any custom code. My only concern would be that we are forcing the use of `#new` with just one parameter and not making it very flexible. So if I wanted to initialize my presenter from many different sources, I'd like to use something like a Factory Method like:

```ruby
class PostPresenter
  def self.wrap(post, *_)
    self.for_admin(post)
  end

  def self.for_admin(post)
    # ...
  end
end
```

But after writing that code I guess what I'd rather do is an `AdminPostPresenter` class instead.

In short, I agree with the change. I'll implement it and change the docs accordingly if that's alright.

Comment (deivid-rodriguez) [reply docs/11-decorators.md:61]: Right.

Maybe instead of explicitly checking for `draper`, we could use `decorator.collection_decorator_class` if it's defined. As long as we document that expected interface `.decorate(<collection>, with: <decorator_class>)`, people should be able to override that to fit their needs.

Comment (deivid-rodriguez) [reply docs/11-decorators.md:61]: Yes, go ahead!

Oh, I forgot I also needed to add `, to: :object` add the end of the `Wrapper.wrap` delegations so that it uses the standard Rails delegation instead of the `draper` shortcut.

Comment (brunvez) [reply docs/11-decorators.md:61]: Yeah, no worries I got that 😄 . BTW with this change, we wouldn't need to delegate or call those methods at all. Since wrapping would be done on the render level all we need are the array methods to iterate. Though I think small changes are better

Comment (deivid-rodriguez) [reply docs/11-decorators.md:80]: My idea is to keep the current wrapping at the same level, and ideally just touch code paths that don't affect the `draper` integration at all. So ideally revert all changes at `lib/active_admin/resource_controller.rb`, `lib/active_admin/resource_controller/data_access.rb`, `lib/active_admin/views/pages/index.rb`, and restore the removed specs.

Maybe delaying the decoration is a good thing, but I'd like to minimize the chance of regressions because of the addition of this feature.

Comment (brunvez) [reply docs/11-decorators.md]: Agreed, I made the proposed changes. Let me know if you have any other comments, happy to implement them.

Comment (deivid-rodriguez) [reply docs/11-decorators.md]: Thanks for your work! I'll do another review pass tomorrow, but should be essentially ready for my side :)

File: docs/11-decorators.md [hunk 3]
```diff
@@ -42,6 +37,40 @@ ActiveAdmin.register Post do
...
+  attr_reader :post
+  delegate_missing_to :post
+
+  def initialize(post)
+    @post = post
+  end
+end
+```
+
+You can also customize the way a collection is wrapped by implementing
```
Comment (deivid-rodriguez) [inline docs/11-decorators.md:55]: After a second thought, and although I believe using duck typing rather that checking from draper directly is fine, I think I wouldn't document this unless there's a good use case for it. The one you mentioned in https://github.com/activeadmin/activeadmin/pull/6249#discussion_r449672211 is probably best implemented as a separate decorator and our decorator support requires decorators to be instantiated by passing the decorated object as the single parameter anyways.

File: lib/active_admin/resource_controller/decorators.rb [hunk 4]
```diff
@@ -48,18 +48,45 @@ def collection_decorator
...
-            name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
-            @cache[name] ||= wrap! collection_decorator, name
-          else
-            decorator
-          end
+          collection_decorator = find_collection_decorator(decorator)
+          name = if collection_decorator <= CollectionDecorator
+                   "#{collection_decorator.name} of #{decorator}"
+                 else
+                   "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource_controller/decorators.rb:87]: Are the two branches here needed? I guess the `+ ActiveAdmin` part makes sense either way since we're adding an extra layer on top of the raw collection decorator.

File: lib/active_admin/resource_controller/decorators.rb [hunk 5]
```diff
@@ -48,18 +48,45 @@ def collection_decorator
         end
       end
 
+      class CollectionDecorator
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource_controller/decorators.rb:51]: This class is essentially copied from Draper, so I would add a commennt on top of it about what it does, and include a permalink to the current version of this class inside Draper for proper attribution.

File: docs/11-decorators.md (hunk 3 above)
Comment (brunvez) [reply docs/11-decorators.md:55]: Agreed, it felt very weird documenting that part

**Comment by brunvez** [comment]
> One final round of comments, but this is awesome and exciting! Could you also add a CHANGELOG entry? Don't forget to add links to PR and your github profile at the bottom :)

More than happy to! I hope I got all comments but lmk if anything else pops into your head

File: lib/active_admin/resource_controller/decorators.rb [hunk 6]
```diff
@@ -48,18 +48,47 @@ def collection_decorator
...
+        delegate :==, :as_json, *array_methods, to: :decorated_collection
+
+        def initialize(object, with:)
+          @object = object
+          @decorator_class = with
+        end
+
+        def self.name
+          'ActiveAdmin::CollectionDecorator'
+        end
```
Comment (deivid-rodriguez) [inline lib/active_admin/resource_controller/decorators.rb:74]: I think it's not a good practice to override `self.name` for non anonymous classes. But I agree that the description looks better with this name. So maybe we can move the class to its own `lib/active_admin/collection_decorator.rb` and nest it directly under `ActiveAdmin` so that it has this better name.

Comment (brunvez) [reply lib/active_admin/resource_controller/decorators.rb:74]: Gladly! Is it alright if I require the new file form `lib/active_admin/resource_controller.rb` or did you have anywhere else in mind?

Comment (deivid-rodriguez) [reply lib/active_admin/resource_controller/decorators.rb:74]: Sure!
